File: whoathere/helpers/macos-vm-helper/Sources/WhoaThereMacosVmHelperCore/LinuxVzTeardownEvidencePayload.swift
import Foundation

public let linuxVzTeardownEvidencePayloadSchemaV1 =
    "whoathere.linux_vz_teardown_evidence_payload.v1"
public let maximumLinuxVzTeardownEvidencePayloadBytesV1 = 64 * 1024
private let linuxVzTeardownEvidencePrefixV1 =
    Data("WHOATHERE_GUEST_TEARDOWN_EVIDENCE ".utf8)

public enum LinuxVzTeardownEvidencePayloadError: Error, Equatable {
    case missing
    case duplicate
    case limitExceeded
    case invalidJSON
    case nonCanonical
    case invalidSchema
    case invalidEvent
}

public struct LinuxVzTeardownEvidencePayload: Equatable, Sendable {
    public let canonicalJSON: Data
    public let payloadSHA256: String
    public let fixtureCase: String
    public let packageUID: UInt64
    public let packageGID: UInt64
    public let claims: LinuxVzTelemetryGuestObservationClaims
}

private struct LinuxVzTeardownEvent {
    let actorPID: UInt64
    let cgroupID: UInt64
    let kind: String
    let sequence: UInt64
    let subjectPID: UInt64
    let timestampNS: UInt64
}

public func decodeLinuxVzTeardownEvidencePayloadV1(
    _ serial: Data
) throws -> LinuxVzTeardownEvidencePayload {
    var payload: Data?
    for rawLine in serial.split(separator: 0x0a, omittingEmptySubsequences: false) {
        var line = Data(rawLine)
        if line.last == 0x0d { line.removeLast() }
        guard line.starts(with: linuxVzTeardownEvidencePrefixV1) else { continue }
        guard payload == nil else { throw LinuxVzTeardownEvidencePayloadError.duplicate }
        payload = line.dropFirst(linuxVzTeardownEvidencePrefixV1.count)
    }
    guard let payload else { throw LinuxVzTeardownEvidencePayloadError.missing }
    return try decodeLinuxVzTeardownEvidenceJSONV1(payload)
}

public func decodeLinuxVzTeardownEvidenceJSONV1(
    _ data: Data
) throws -> LinuxVzTeardownEvidencePayload {
    guard !data.isEmpty, data.count <= maximumLinuxVzTeardownEvidencePayloadBytesV1 else {
        throw LinuxVzTeardownEvidencePayloadError.limitExceeded
    }
    guard let value = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
        throw LinuxVzTeardownEvidencePayloadError.invalidJSON
    }
    guard try canonicalJSONData(value) == data else {
        throw LinuxVzTeardownEvidencePayloadError.nonCanonical
    }
    let baseKeys = Set([
        "cgroup_empty_after_reap", "cgroup_removed", "deadline_limit_ns", "deadline_reached",
        "descendant_teardown_complete", "dropped_event_count", "event_count",
        "event_sequence_end", "event_sequence_start", "events", "evidence_truncated",
        "fixture_case", "heartbeat_count", "kill_signal_count",
        "package_gid", "package_uid", "reaped_process_count", "schema_version",
        "sensor_healthy", "sensor_teardown_complete", "teardown_trigger",
        "termination_signal_count"
    ])
    guard let fixtureCase = value["fixture_case"] as? String else {
        throw LinuxVzTeardownEvidencePayloadError.invalidSchema
    }
    let expectedKeys: Set<String>
    let expectedKinds: [String]
    let observedTerminal: String
    switch fixtureCase {
    case "normal_exit":
        expectedKeys = baseKeys.union(["fixture_exit_status"])
        expectedKinds = ["fork", "exec", "exit"]
        observedTerminal = "observation_complete"
        guard value["teardown_trigger"] as? String == "natural_exit",
              teardownDecimal(value["deadline_limit_ns"]) == 5_000_000_000,
              value["deadline_reached"] as? Bool == false,
              teardownDecimal(value["fixture_exit_status"]) == 0,
              value["fixture_termination_signal"] == nil,
              value["term_grace_limit_ns"] == nil,
              value["term_grace_reached"] == nil,
              value["term_resistance_proven"] == nil,
              teardownDecimal(value["termination_signal_count"]) == 0,
              teardownDecimal(value["kill_signal_count"]) == 0 else {
            throw LinuxVzTeardownEvidencePayloadError.invalidSchema
        }
    case "timeout":
        expectedKeys = baseKeys.union(["fixture_termination_signal"])
        expectedKinds = ["fork", "exec", "signal_term", "exit"]
        observedTerminal = "timeout_with_teardown"
        guard value["teardown_trigger"] as? String == "deadline",
              teardownDecimal(value["deadline_limit_ns"]) == 1_000_000_000,
              value["deadline_reached"] as? Bool == true,
              value["fixture_exit_status"] == nil,
              teardownDecimal(value["fixture_termination_signal"]) == 15,
              value["term_grace_limit_ns"] == nil,
              value["term_grace_reached"] == nil,
              value["term_resistance_proven"] == nil,
              teardownDecimal(value["termination_signal_count"]) == 1,
              teardownDecimal(value["kill_signal_count"]) == 0 else {
            throw LinuxVzTeardownEvidencePayloadError.invalidSchema
        }
    case "term_resistance":
        expectedKeys = baseKeys.union([
            "fixture_termination_signal", "term_grace_limit_ns", "term_grace_reached",
            "term_resistance_proven"
        ])
        expectedKinds = ["fork", "exec", "signal_term", "signal_kill", "exit"]
        observedTerminal = "timeout_with_teardown"
        guard value["teardown_trigger"] as? String == "deadline",
              teardownDecimal(value["deadline_limit_ns"]) == 1_000_000_000,
              value["deadline_reached"] as? Bool == true,
              value["fixture_exit_status"] == nil,
              teardownDecimal(value["fixture_termination_signal"]) == 9,
              teardownDecimal(value["term_grace_limit_ns"]) == 250_000_000,
              value["term_grace_reached"] as? Bool == true,
              value["term_resistance_proven"] as? Bool == true,
              teardownDecimal(value["termination_signal_count"]) == 1,
              teardownDecimal(value["kill_signal_count"]) == 1 else {
            throw LinuxVzTeardownEvidencePayloadError.invalidSchema
        }
    case "escaped_session":
        expectedKeys = baseKeys.union([
            "fixture_termination_signal", "session_escape_count", "session_target"
        ])
        expectedKinds = ["fork", "exec", "setsid", "signal_term", "exit"]
        observedTerminal = "timeout_with_teardown"
        guard value["teardown_trigger"] as? String == "deadline",
              teardownDecimal(value["deadline_limit_ns"]) == 1_000_000_000,
              value["deadline_reached"] as? Bool == true,
              value["fixture_exit_status"] == nil,
              teardownDecimal(value["fixture_termination_signal"]) == 15,
              teardownDecimal(value["session_escape_count"]) == 1,
              value["session_target"] as? String == "new_session_leader_at_deadline",
              value["term_grace_limit_ns"] == nil,
              value["term_grace_reached"] == nil,
              value["term_resistance_proven"] == nil,
              teardownDecimal(value["termination_signal_count"]) == 1,
              teardownDecimal(value["kill_signal_count"]) == 0 else {
            throw LinuxVzTeardownEvidencePayloadError.invalidSchema
        }
    case "reparented_child":
        expectedKeys = baseKeys.union([
            "fixture_termination_signal", "fork_count", "reparent_target",
            "reparented_process_count"
        ])
        expectedKinds = ["exec", "fork", "reparent", "signal_term", "exit"]
        observedTerminal = "timeout_with_teardown"
        guard value["teardown_trigger"] as? String == "deadline",
              teardownDecimal(value["deadline_limit_ns"]) == 1_000_000_000,
              value["deadline_reached"] as? Bool == true,
              value["fixture_exit_status"] == nil,
              teardownDecimal(value["fixture_termination_signal"]) == 15,
              teardownDecimal(value["fork_count"]) == 2,
              teardownDecimal(value["reaped_process_count"]) == 2,
              value["reparent_target"] as? String == "protected_subreaper_at_deadline",
              teardownDecimal(value["reparented_process_count"]) == 1,
              value["session_escape_count"] == nil,
              value["session_target"] == nil,
              value["term_grace_limit_ns"] == nil,
              value["term_grace_reached"] == nil,
              value["term_resistance_proven"] == nil,
              teardownDecimal(value["termination_signal_count"]) == 1,
              teardownDecimal(value["kill_signal_count"]) == 0 else {
            throw LinuxVzTeardownEvidencePayloadError.invalidSchema
        }
    default:
        throw LinuxVzTeardownEvidencePayloadError.invalidSchema
    }
    let eventCount = UInt64(expectedKinds.count)
    guard Set(value.keys) == expectedKeys,
          value["schema_version"] as? String == linuxVzTeardownEvidencePayloadSchemaV1,
          value["descendant_teardown_complete"] as? Bool == true,
          value["cgroup_empty_after_reap"] as? Bool == true,
          value["cgroup_removed"] as? Bool == true,
          value["sensor_teardown_complete"] as? Bool == true,
          value["sensor_healthy"] as? Bool == true,
          value["evidence_truncated"] as? Bool == false,
          teardownDecimal(value["reaped_process_count"]) ==
            (fixtureCase == "reparented_child" ? 2 : 1),
          teardownDecimal(value["event_sequence_start"]) == 1,
          teardownDecimal(value["event_sequence_end"]) == eventCount,
          teardownDecimal(value["event_count"]) == eventCount,
          teardownDecimal(value["heartbeat_count"]) == 2,
          teardownDecimal(value["dropped_event_count"]) == 0,
          teardownDecimal(value["package_uid"]) == 65534,
          teardownDecimal(value["package_gid"]) == 65534,
          let rawEvents = value["events"] as? [[String: Any]],
          rawEvents.count == expectedKinds.count else {
        throw LinuxVzTeardownEvidencePayloadError.invalidSchema
    }
    let events = try rawEvents.map(decodeLinuxVzTeardownEvent)
    guard events.map(\.kind) == expectedKinds,
          events.enumerated().allSatisfy({ $0.element.sequence == UInt64($0.offset + 1) }),
          events.allSatisfy({
              $0.actorPID > 0 && $0.subjectPID > 0 && $0.cgroupID > 0 && $0.timestampNS > 0
          }),
          events.dropFirst().allSatisfy({ $0.cgroupID == events[0].cgroupID }),
          zip(events, events.dropFirst()).allSatisfy({ pair in
              pair.0.timestampNS < pair.1.timestampNS
          }),
          fixtureCase == "reparented_child" ||
            events[0].actorPID != events[0].subjectPID else {
        throw LinuxVzTeardownEvidencePayloadError.invalidEvent
    }
    switch fixtureCase {
    case "normal_exit":
        guard events.dropFirst().allSatisfy({
            $0.actorPID == events[0].subjectPID && $0.subjectPID == events[0].subjectPID
        }) else {
            throw LinuxVzTeardownEvidencePayloadError.invalidEvent
        }
    case "timeout":
        guard events[1].actorPID == events[0].subjectPID,
              events[1].subjectPID == events[0].subjectPID,
              events[2].actorPID == events[0].actorPID,
              events[2].subjectPID == events[0].subjectPID,
              events[3].actorPID == events[0].subjectPID,
              events[3].subjectPID == events[0].subjectPID else {
            throw LinuxVzTeardownEvidencePayloadError.invalidEvent
        }
    case "term_resistance":
        guard events[1].actorPID == events[0].subjectPID,
              events[1].subjectPID == events[0].subjectPID,
              events[2].actorPID == events[0].actorPID,
              events[2].subjectPID == events[0].subjectPID,
              events[3].actorPID == events[0].actorPID,
              events[3].subjectPID == events[0].subjectPID,
              events[4].actorPID == events[0].subjectPID,
              events[4].subjectPID == events[0].subjectPID else {
            throw LinuxVzTeardownEvidencePayloadError.invalidEvent
        }
    case "escaped_session":
        guard events[1].actorPID == events[0].subjectPID,
              events[1].subjectPID == events[0].subjectPID,
              events[2].actorPID == events[0].subjectPID,
              events[2].subjectPID == events[0].subjectPID,
              events[3].actorPID == events[0].actorPID,
              events[3].subjectPID == events[0].subjectPID,
              events[4].actorPID == events[0].subjectPID,
              events[4].subjectPID == events[0].subjectPID else {
            throw LinuxVzTeardownEvidencePayloadError.invalidEvent
        }
    case "reparented_child":
        guard events[0].actorPID == events[0].subjectPID,
              events[1].actorPID == events[0].actorPID,
              events[1].subjectPID != events[0].actorPID,
              events[2].actorPID != events[0].actorPID,
              events[2].actorPID != events[1].subjectPID,
              events[2].subjectPID == events[1].subjectPID,
              events[3].actorPID == events[2].actorPID,
              events[3].subjectPID == events[1].subjectPID,
              events[4].actorPID == events[1].subjectPID,
              events[4].subjectPID == events[1].subjectPID else {
            throw LinuxVzTeardownEvidencePayloadError.invalidEvent
        }
    default:
        throw LinuxVzTeardownEvidencePayloadError.invalidSchema
    }
    let claims = LinuxVzTelemetryGuestObservationClaims(
        evidencePayloadSHA256: sha256(data),
        evidenceByteLength: UInt64(data.count),
        eventSequenceStart: 1,
        eventSequenceEnd: eventCount,
        eventCount: eventCount,
        heartbeatCount: 2,
        droppedEventCount: 0,
        sensorHealthy: true,
        evidenceTruncated: false,
        descendantTeardownComplete: true,
        observedTerminal: observedTerminal
    )
    return LinuxVzTeardownEvidencePayload(
        canonicalJSON: data,
        payloadSHA256: claims.evidencePayloadSHA256,
        fixtureCase: fixtureCase,
        packageUID: 65534,
        packageGID: 65534,
        claims: claims
    )
}

private func decodeLinuxVzTeardownEvent(
    _ value: [String: Any]
) throws -> LinuxVzTeardownEvent {
    guard Set(value.keys) == Set([
        "actor_pid", "cgroup_id", "kind", "sequence", "subject_pid", "timestamp_ns"
    ]),
    let actorPID = teardownDecimal(value["actor_pid"]),
    let cgroupID = teardownDecimal(value["cgroup_id"]),
    let kind = value["kind"] as? String,
    let sequence = teardownDecimal(value["sequence"]),
    let subjectPID = teardownDecimal(value["subject_pid"]),
    let timestampNS = teardownDecimal(value["timestamp_ns"]) else {
        throw LinuxVzTeardownEvidencePayloadError.invalidEvent
    }
    return LinuxVzTeardownEvent(
        actorPID: actorPID,
        cgroupID: cgroupID,
        kind: kind,
        sequence: sequence,
        subjectPID: subjectPID,
        timestampNS: timestampNS
    )
}

private func teardownDecimal(_ value: Any?) -> UInt64? {
    guard let text = value as? String, !text.isEmpty,
          text == "0" || (text.first != "0" && text.allSatisfy(\.isNumber)) else {
        return nil
    }
    return UInt64(text)
}

File: whoathere/crates/whoathere-macos-vm/src/linux_vz_teardown_evidence_payload.rs
use crate::{
    LinuxVzTelemetryConformanceCaseV1, LinuxVzTelemetryConformanceObservedTerminalV1,
    LinuxVzTelemetryGuestObservationClaimsV1, MacosLinuxVzTelemetryEvidenceErrorV1,
};
use serde::{Deserialize, Serialize};
use std::fmt;
use whoathere_artifact::Sha256Digest;

pub const LINUX_VZ_TEARDOWN_EVIDENCE_PAYLOAD_SCHEMA_V1: &str =
    "whoathere.linux_vz_teardown_evidence_payload.v1";
pub const LINUX_VZ_TEARDOWN_EVIDENCE_SERIAL_PREFIX_V1: &[u8] =
    b"WHOATHERE_GUEST_TEARDOWN_EVIDENCE ";
pub const MAX_LINUX_VZ_TEARDOWN_EVIDENCE_PAYLOAD_BYTES_V1: usize = 64 * 1024;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum LinuxVzTeardownEvidencePayloadErrorV1 {
    Missing,
    Duplicate,
    LimitExceeded,
    InvalidJson,
    NonCanonical,
    InvalidSchema,
    InvalidEvent,
}

impl fmt::Display for LinuxVzTeardownEvidencePayloadErrorV1 {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(match self {
            Self::Missing => "linux_vz_teardown_evidence_missing",
            Self::Duplicate => "linux_vz_teardown_evidence_duplicate",
            Self::LimitExceeded => "linux_vz_teardown_evidence_limit_exceeded",
            Self::InvalidJson => "linux_vz_teardown_evidence_json_invalid",
            Self::NonCanonical => "linux_vz_teardown_evidence_noncanonical",
            Self::InvalidSchema => "linux_vz_teardown_evidence_schema_invalid",
            Self::InvalidEvent => "linux_vz_teardown_evidence_event_invalid",
        })
    }
}

impl std::error::Error for LinuxVzTeardownEvidencePayloadErrorV1 {}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct LinuxVzTeardownEvidencePayloadV1 {
    canonical_json: Vec<u8>,
    payload_sha256: Sha256Digest,
}

impl LinuxVzTeardownEvidencePayloadV1 {
    pub fn canonical_json_v1(&self) -> &[u8] {
        &self.canonical_json
    }

    pub fn payload_sha256(&self) -> &Sha256Digest {
        &self.payload_sha256
    }

    pub fn fixture_case(&self) -> LinuxVzTelemetryConformanceCaseV1 {
        LinuxVzTelemetryConformanceCaseV1::NormalExit
    }

    pub fn package_uid(&self) -> u32 {
        65534
    }

    pub fn package_gid(&self) -> u32 {
        65534
    }

    pub fn guest_observation_claims_v1(
        &self,
    ) -> Result<LinuxVzTelemetryGuestObservationClaimsV1, MacosLinuxVzTelemetryEvidenceErrorV1>
    {
        LinuxVzTelemetryGuestObservationClaimsV1::new(
            self.payload_sha256.clone(),
            self.canonical_json.len() as u64,
            1,
            3,
            3,
            2,
            0,
            true,
            false,
            true,
            LinuxVzTelemetryConformanceObservedTerminalV1::ObservationComplete,
        )
    }
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
struct TeardownEvidenceWireV1 {
    cgroup_empty_after_reap: bool,
    cgroup_removed: bool,
    deadline_limit_ns: String,
    deadline_reached: bool,
    descendant_teardown_complete: bool,
    dropped_event_count: String,
    event_count: String,
    event_sequence_end: String,
    event_sequence_start: String,
    events: Vec<TeardownEventWireV1>,
    evidence_truncated: bool,
    fixture_case: String,
    fixture_exit_status: String,
    heartbeat_count: String,
    kill_signal_count: String,
    package_gid: String,
    package_uid: String,
    reaped_process_count: String,
    schema_version: String,
    sensor_healthy: bool,
    sensor_teardown_complete: bool,
    teardown_trigger: String,
    termination_signal_count: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
struct TeardownEventWireV1 {
    actor_pid: String,
    cgroup_id: String,
    kind: String,
    sequence: String,
    subject_pid: String,
    timestamp_ns: String,
}

pub fn decode_linux_vz_teardown_evidence_from_serial_v1(
    serial: &[u8],
) -> Result<LinuxVzTeardownEvidencePayloadV1, LinuxVzTeardownEvidencePayloadErrorV1> {
    let mut payload = None;
    for raw_line in serial.split(|byte| *byte == b'\n') {
        let line = raw_line.strip_suffix(b"\r").unwrap_or(raw_line);
        let Some(candidate) = line.strip_prefix(LINUX_VZ_TEARDOWN_EVIDENCE_SERIAL_PREFIX_V1) else {
            continue;
        };
        if payload.replace(candidate).is_some() {
            return Err(LinuxVzTeardownEvidencePayloadErrorV1::Duplicate);
        }
    }
    decode_linux_vz_teardown_evidence_payload_v1(
        payload.ok_or(LinuxVzTeardownEvidencePayloadErrorV1::Missing)?,
    )
}

pub fn decode_linux_vz_teardown_evidence_payload_v1(
    payload: &[u8],
) -> Result<LinuxVzTeardownEvidencePayloadV1, LinuxVzTeardownEvidencePayloadErrorV1> {
    if payload.is_empty() || payload.len() > MAX_LINUX_VZ_TEARDOWN_EVIDENCE_PAYLOAD_BYTES_V1 {
        return Err(LinuxVzTeardownEvidencePayloadErrorV1::LimitExceeded);
    }
    let mut deserializer = serde_json::Deserializer::from_slice(payload);
    let wire = TeardownEvidenceWireV1::deserialize(&mut deserializer)
        .map_err(|_| LinuxVzTeardownEvidencePayloadErrorV1::InvalidJson)?;
    deserializer
        .end()
        .map_err(|_| LinuxVzTeardownEvidencePayloadErrorV1::InvalidJson)?;
    let canonical = serde_json_canonicalizer::to_vec(&wire)
        .map_err(|_| LinuxVzTeardownEvidencePayloadErrorV1::InvalidJson)?;
    if canonical != payload {
        return Err(LinuxVzTeardownEvidencePayloadErrorV1::NonCanonical);
    }

    if wire.schema_version != LINUX_VZ_TEARDOWN_EVIDENCE_PAYLOAD_SCHEMA_V1
        || wire.fixture_case != "normal_exit"
        || wire.teardown_trigger != "natural_exit"
        || decimal_u64_v1(&wire.deadline_limit_ns)? != 5_000_000_000
        || wire.deadline_reached
        || !wire.descendant_teardown_complete
        || !wire.cgroup_empty_after_reap
        || !wire.cgroup_removed
        || !wire.sensor_teardown_complete
        || !wire.sensor_healthy
        || wire.evidence_truncated
        || decimal_u64_v1(&wire.fixture_exit_status)? != 0
        || decimal_u64_v1(&wire.termination_signal_count)? != 0
        || decimal_u64_v1(&wire.kill_signal_count)? != 0
        || decimal_u64_v1(&wire.reaped_process_count)? != 1
        || decimal_u64_v1(&wire.event_sequence_start)? != 1
        || decimal_u64_v1(&wire.event_sequence_end)? != 3
        || decimal_u64_v1(&wire.event_count)? != 3
        || decimal_u64_v1(&wire.heartbeat_count)? != 2
        || decimal_u64_v1(&wire.dropped_event_count)? != 0
        || decimal_u64_v1(&wire.package_uid)? != 65534
        || decimal_u64_v1(&wire.package_gid)? != 65534
        || wire.events.len() != 3
    {
        return Err(LinuxVzTeardownEvidencePayloadErrorV1::InvalidSchema);
    }

    let mut actor_pids = [0_u64; 3];
    let mut subject_pids = [0_u64; 3];
    let mut cgroup_ids = [0_u64; 3];
    let mut timestamps = [0_u64; 3];
    for (index, event) in wire.events.iter().enumerate() {
        if event.kind != ["fork", "exec", "exit"][index]
            || decimal_u64_v1(&event.sequence)? != (index + 1) as u64
        {
            return Err(LinuxVzTeardownEvidencePayloadErrorV1::InvalidEvent);
        }
        actor_pids[index] = decimal_u64_v1(&event.actor_pid)?;
        subject_pids[index] = decimal_u64_v1(&event.subject_pid)?;
        cgroup_ids[index] = decimal_u64_v1(&event.cgroup_id)?;
        timestamps[index] = decimal_u64_v1(&event.timestamp_ns)?;
    }
    if actor_pids.contains(&0)
        || subject_pids.contains(&0)
        || cgroup_ids.contains(&0)
        || timestamps.contains(&0)
        || actor_pids[0] == subject_pids[0]
        || actor_pids[1] != subject_pids[0]
        || subject_pids[1] != subject_pids[0]
        || actor_pids[2] != subject_pids[0]
        || subject_pids[2] != subject_pids[0]
        || cgroup_ids[1] != cgroup_ids[0]
        || cgroup_ids[2] != cgroup_ids[0]
        || timestamps[0] >= timestamps[1]
        || timestamps[1] >= timestamps[2]
    {
        return Err(LinuxVzTeardownEvidencePayloadErrorV1::InvalidEvent);
    }

    Ok(LinuxVzTeardownEvidencePayloadV1 {
        canonical_json: canonical,
        payload_sha256: Sha256Digest::from_bytes(payload),
    })
}

fn decimal_u64_v1(value: &str) -> Result<u64, LinuxVzTeardownEvidencePayloadErrorV1> {
    if value.is_empty()
        || (value != "0" && value.starts_with('0'))
        || !value.bytes().all(|byte| byte.is_ascii_digit())
    {
        return Err(LinuxVzTeardownEvidencePayloadErrorV1::InvalidSchema);
    }
    value
        .parse()
        .map_err(|_| LinuxVzTeardownEvidencePayloadErrorV1::InvalidSchema)
}

#[cfg(test)]
mod tests {
    use super::*;

    const NORMAL_EXIT: &[u8] = br#"{"cgroup_empty_after_reap":true,"cgroup_removed":true,"deadline_limit_ns":"5000000000","deadline_reached":false,"descendant_teardown_complete":true,"dropped_event_count":"0","event_count":"3","event_sequence_end":"3","event_sequence_start":"1","events":[{"actor_pid":"41","cgroup_id":"9001","kind":"fork","sequence":"1","subject_pid":"42","timestamp_ns":"100"},{"actor_pid":"42","cgroup_id":"9001","kind":"exec","sequence":"2","subject_pid":"42","timestamp_ns":"200"},{"actor_pid":"42","cgroup_id":"9001","kind":"exit","sequence":"3","subject_pid":"42","timestamp_ns":"300"}],"evidence_truncated":false,"fixture_case":"normal_exit","fixture_exit_status":"0","heartbeat_count":"2","kill_signal_count":"0","package_gid":"65534","package_uid":"65534","reaped_process_count":"1","schema_version":"whoathere.linux_vz_teardown_evidence_payload.v1","sensor_healthy":true,"sensor_teardown_complete":true,"teardown_trigger":"natural_exit","termination_signal_count":"0"}"#;

    #[test]
    fn normal_exit_binds_natural_exit_lineage_and_complete_cleanup() {
        let payload = decode_linux_vz_teardown_evidence_payload_v1(NORMAL_EXIT).unwrap();
        assert_eq!(
            payload.fixture_case(),
            LinuxVzTelemetryConformanceCaseV1::NormalExit
        );
        assert_eq!(payload.package_uid(), 65534);
        let claims = payload.guest_observation_claims_v1().unwrap();
        assert_eq!(claims.dropped_event_count(), 0);
        assert!(claims.sensor_healthy());
        assert!(claims.descendant_teardown_complete());
    }

    #[test]
    fn normal_exit_rejects_forged_cleanup_signal_and_lineage_claims() {
        for changed in [
            String::from_utf8(NORMAL_EXIT.to_vec())
                .unwrap()
                .replace("\"cgroup_removed\":true", "\"cgroup_removed\":false"),
            String::from_utf8(NORMAL_EXIT.to_vec())
                .unwrap()
                .replace("\"deadline_reached\":false", "\"deadline_reached\":true"),
            String::from_utf8(NORMAL_EXIT.to_vec()).unwrap().replace(
                "\"termination_signal_count\":\"0\"",
                "\"termination_signal_count\":\"1\"",
            ),
            String::from_utf8(NORMAL_EXIT.to_vec())
                .unwrap()
                .replace("\"timestamp_ns\":\"200\"", "\"timestamp_ns\":\"99\""),
        ] {
            assert!(decode_linux_vz_teardown_evidence_payload_v1(changed.as_bytes()).is_err());
        }
    }
}
